enhancements/dns/coredns-egressfirewall-integration.md

```diff
@@ -0,0 +1,320 @@
+allow/deny access to all the subdomains belonging to the wildcard DNS names.
+* As an OpenShift engineer, I want to add a new Custom Resource, so that IPs and TTLs of the DNS names which are used
+in EgressFirewall rules can be tracked.
+* As an OpenShift engineer, I want to add a new plugin to CoreDNS, so that DNS lookups of DNS names used in EgressFirewall
+rules can be inspected and the current IPs and TTLs can be tracked in the corresponding new CR.
+* As an OpenShift engineer, I want to modify Cluster DNS operator, so that CoreDNS can be deployed with the
+new plugin enabled.
+
+### Goals
+
+* Support update of the ``AddressSet`` referenced by the ACL rule corresponding to an EgressFirewall if IPs associated
+with the corresponding DNS names change dynamically.
+* Support usage of wildcard DNS names in EgressFirewall rules.
+* Create CRs for each unique DNS name used in the EgressFirewall rules and use the CRs to track the current
+IPs and the corresponding TTL information.
+
+### Non-Goals
+
+* Support additional DNS resolution functionality in the new CoreDNS plugin.
+* Support allowing/denying of DNS lookups (core EgressFirewall functionality) in the new CoreDNS plugin.
+
+## Proposal
+
+This enhancement proposes to introduce a new CoreDNS plugin (``egressfirewall``) and a new Custom Resource
+(``DNSName``) to improve the integration of CoreDNS with EgressFirewall. The ``DNSName`` CR will be created
+for each unique DNS name (both regular and wildcard DNS names) used in the EgressFirewall rules. This CR will
+be used to store the DNS name along with the current IPs and the correspodning TTL and the next lookup time
+based on the TTL. The OVN-K master will be responsible for the creation of the ``DNSName`` CRs. The ``DNSName``
+CR is meant for communication between CoreDNS and OVN-K master. Thus it cannot be created or modified by an
+OpenShift cluster adminisrator.
+
+The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
+plugins. If the DNS name in the query matches any ``DNSName`` CR(s) (regular or wildcard or both), then the
+plugin will update the ``.status`` of the matching ``DNSName`` CR(s) with the DNS name along with the IPs and
+the corresponding TTL and the next lookup time based on the TTL. The OVN-K master will watch the ``DNSName``
+CRs. Whenever the IPs are updated for a ``DNSName`` CR, the OVN-K master will update the the underlying ``AdressSet``
+referenced by the ACL rule(s) for the corresponding EgressFirewall(s).
+
+OVN-K master will keep track of the TTL (or next lookup time) for each regular DNS name and send a DNS lookup
+query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed
+directly on the DNS name as it will not return any IP. Thus, the lookups will be performed on the DNS names
+which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.
```

Comment:
+1, another question - if I got that right, to enable frequent IP updates, coreDNS needs feedback from the ovn-k master, to ensure ip changes are applied for egress firewall before pods start using them.
So the status may need to include a filed for that feedback from ovn-k master?

Reply:
Maybe a "generation" ID in the CR that ovnkube updates? Whenever that changes CoreDNS would know that set of updates was applied.

Reply:
OVN-K not re-resolving the IPs whenever the TTL expires has some downsides:
1. This would mean that the additional delay introduced in the DNS lookup process for updating the new IPs in the underlying ACL will always be experienced by the client pods. If OVN-K continues to do the re-resolving of the IPs, according to the current behavior, the delay in DNS lookup will mostly be faced by OVN-K and not by the client pods. Thus the overall user experience is not degraded.
2. For wildcard DNS names, the individual subdomains are added to the ``.status`` of the ``DNSName`` CR. Now once these are added OVN-K must check whether these have become stale or not. Thus a retry counter is proposed for every DNS name which will be tracked by OVN-K. If the retry counter crosses a threshold value, then the subdomain will be removed from the ``.status`` of the ``DNSName``. If this is not done then egressfirewall will incorrectly allow certain IPs.

Reply:
The CoreDNS pods run as Daemonsets and they run independent of each other. So if the CoreDNS tracks the TTL and does the re-resolution, then all the CoreDNS pods would be doing the same work and there'll be chances of race condition while updating the IPs in the ``.status`` of the ``DNSName`` CRs.

Reply:
How is it different from the `DNSName` creation? I mean if ovn-k master creates a new `DNSName` object, some CoreDNS pod needs to resolve it right? Which CoreDNS pod will do that?

Reply:
The DNS lookup request is sent to the CoreDNS service and whichever CoreDNS pod receives the DNS lookup query from OVN-K will respond to the query. If the re-resolution request is sent by the CoreDNS pods, then all of them will raise the same request as all of them are independent of each other.

Reply:
oh i missed that part where ovn-k needs to send the request together with creating the `DNSName` object. so for wildcarded objects it should set random string instead of `*` to trigger resolution?
In general, sounds like a very tight integration between ovn-k and `CoreDNS` implementation, and as @danwinship said that may not match the initial goal of this feature?

Reply:
@npinaeva the process for wildcard DNS names is explained here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R129-R150

Reply:
The goal of the feature is to update the underlying ACLs for EgressFirewall first before CoreDNS responds to pods with the DNS lookup response so that pods are not incorrectly allowed/denied access to DNS names. Another goal is to support wildcard DNS names. These goals can only be supported with a better integration between EgressFirewall and CoreDNS.

Currently, the defaultTTL for re-resolving a DNS name is set to 30s in OVN-K. We can tinker with this value and set it to a higher value so that the re-resolution is done less frequently. However, as I mentioned in the [downsides](https://github.com/openshift/enhancements/pull/1335#discussion_r1107544210) of removing this existing behavior from OVN-K, both CoreDNS and EgressFirewall will not behave as expected.

Reply:
Hm... I hadn't been considering the fact that there are many CoreDNS services, not just one. That makes communication via API objects much trickier...

My original idea in jira had been that CoreDNS and ovn-k would communicate via gRPC, not via the apiserver. But you still need a global view of allowed/denied IPs, because pods might end up talking to a CoreDNS server on a different node during updates when their own local DNS server is unavailable. Which means this is incompatible with OVN-IC (unless every CoreDNS plugin communicates with _every_ ovnkube-master, which seems... very complicated).

So... the plugin will have to deal with merging its response with the existing `status`... which I guess means that we _do_ need to keep the TTL data in the status, so that a plugin can distinguish "old IP that can be removed" vs "IP that another node saw and I didn't, but which is still valid".